### PR TITLE
CIF-2195 - Add additional recommendation types

### PR DESCRIPTION
**Pull request by herzog31**
## Description

* Updated `@magento/recommendations-js-sdk` to version 2.0.5 to support additional recommendation types.
* The recommendation types that can be selected in the dialog of the product recommendation component are now controlled via a template policy. The types that are allowed for a specific template can be checked/unchecked and the dialog of the component uses the new `AllowedRecTypesDataSourceServlet` to retrieve a list of enabled types.
* This allows to set different recommendation types for product and regular CMS pages (as required by Commerce).
* Sample policies are added via https://github.com/adobe/aem-cif-guides-venia/pull/220.

## How Has This Been Tested?

* Unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes and the overall coverage did not decrease.
- [X] All unit tests pass on CircleCi.
- [X] I ran all tests locally and they pass.

### CODE REVIEW

File: extensions/product-recs/bundle/src/main/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServlet.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    /**
+     * Returns a list of allowed product recommendation types based on the template policy.
+     */
+    private List<Resource> getAllowedRecTypes(@NotNull SlingHttpServletRequest request) {
+        List<Resource> allowedRecTypes = new ArrayList<>();
+
+        ResourceResolver resolver = request.getResourceResolver();
+        Resource contentResource = resolver.getResource((String) request.getAttribute(Value.CONTENTPATH_ATTRIBUTE));
+        ContentPolicyManager policyManager = resolver.adaptTo(ContentPolicyManager.class);
+        if (contentResource != null && policyManager != null) {
```
Comment (Buuhuu) [inline extensions/product-recs/bundle/src/main/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServlet.java:72]: You may want to use Stream/Optional APIs here to keep the nested depth low(er)

File: extensions/product-recs/bundle/src/main/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServlet.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            ContentPolicy policy = policyManager.getPolicy(contentResource);
+            if (policy != null) {
+                ValueMap props = policy.getProperties();
+                if (props != null) {
+                    // String[] headingElements = props.get(PN_ALLOWED_HEADING_ELEMENTS, String[].class);
```
Comment (Buuhuu) [inline extensions/product-recs/bundle/src/main/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServlet.java:77]: can we remove the commented code?

File: extensions/product-recs/bundle/src/main/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServlet.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        public String getValue() {
+            return recType.getValue();
+        }
+    }
+
+    /**
+     * Enum with all available product recommendation types.
+     */
+    enum RecType {
+        mostViewed("most-viewed", "Most viewed"),
```
Comment (Buuhuu) [inline extensions/product-recs/bundle/src/main/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServlet.java:140]: please follow the [naming convention of constants in java](https://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html#d5e8254) and make the enum values it all uppercase separated with underscores.

File: extensions/product-recs/bundle/src/test/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServletTest.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+public class AllowedRecTypesDataSourceServletTest {
+
+    private AllowedRecTypesDataSourceServlet dataSourceServlet;
+
+    @Rule
+    public final AemContext context = createContext("/context/datasource-content.json");
```
Comment (Buuhuu) [inline extensions/product-recs/bundle/src/test/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServletTest.java:43]: Please use the `TestContext.newAemContext("/context/datasource-content.json")` factory instead.

Also using context initialiser callbacks is meant to be used by plugins. Tests should use @Before methods instead.

File: extensions/product-recs/content/src/main/content/jcr_root/apps/core/cif/extensions/product-recs/components/productrecommendations/v1/productrecommendations/_cq_design_dialog/.content.xml [hunk 5]
```diff
@@ -2,5 +2,133 @@
...
+                                        jcr:primaryType="nt:unstructured"
+                                        sling:resourceType="granite/ui/components/coral/foundation/text"
+                                        text="Allowed Recommendation Types for Authors"/>
+                                    <field
+                                        cq:showOnCreate="{Boolean}true"
+                                        jcr:primaryType="nt:unstructured"
+                                        sling:resourceType="granite/ui/components/coral/foundation/form/fieldset"
+                                        fieldDescription="List of product recommendation types for authors to choose from.">
+                                        <items jcr:primaryType="nt:unstructured">
+                                            <mostViewed
```
Comment (Buuhuu) [inline extensions/product-recs/content/src/main/content/jcr_root/apps/core/cif/extensions/product-recs/components/productrecommendations/v1/productrecommendations/_cq_design_dialog/.content.xml:36]: Wouldn't a select box with multiselect work better here? I would take less space in the dialog at least.

File: extensions/product-recs/bundle/src/main/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServlet.java (hunk 1 above)
Comment (herzog31) [reply extensions/product-recs/bundle/src/main/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServlet.java]: Done.

File: extensions/product-recs/bundle/src/main/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServlet.java (hunk 3 above)
Comment (herzog31) [reply extensions/product-recs/bundle/src/main/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServlet.java]: Done.

File: extensions/product-recs/bundle/src/test/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServletTest.java (hunk 4 above)
Comment (herzog31) [reply extensions/product-recs/bundle/src/test/java/com/adobe/cq/commerce/extensions/recommendations/internal/servlets/AllowedRecTypesDataSourceServletTest.java]: Makes sense. The class however is not exported from the core bundle. So I copied it in the extension bundle.

File: extensions/product-recs/content/src/main/content/jcr_root/apps/core/cif/extensions/product-recs/components/productrecommendations/v1/productrecommendations/_cq_design_dialog/.content.xml (hunk 5 above)
Comment (herzog31) [reply extensions/product-recs/content/src/main/content/jcr_root/apps/core/cif/extensions/product-recs/components/productrecommendations/v1/productrecommendations/_cq_design_dialog/.content.xml:36]: Will have a look 👍

**Review by Buuhuu — APPROVED** [review]

File: extensions/product-recs/content/src/main/content/jcr_root/apps/core/cif/extensions/product-recs/components/productrecommendations/v1/productrecommendations/_cq_design_dialog/.content.xml (hunk 5 above)
Comment (herzog31) [reply extensions/product-recs/content/src/main/content/jcr_root/apps/core/cif/extensions/product-recs/components/productrecommendations/v1/productrecommendations/_cq_design_dialog/.content.xml:34]: Done.

**Pull request merged**